source/extensions/filters/network/mongo_proxy/codec.h

```diff
@@ -23,15 +23,15 @@ class Message {
 public:
   enum class OpCode {
     OP_REPLY = 1,
-    OP_MSG = 1000,
     OP_UPDATE = 2001,
     OP_INSERT = 2002,
     OP_QUERY = 2004,
     OP_GET_MORE = 2005,
     OP_DELETE = 2006,
     OP_KILL_CURSORS = 2007,
     OP_COMMAND = 2010,
-    OP_COMMANDREPLY = 2011
+    OP_COMMANDREPLY = 2011,
+    OP_MSG = 2013
```

Comment:
for educational purposes, what is the difference between the existing op_msg and the new one?

Reply:
Sorry for the review delay. @oneleggedredcow before I dig into the details, I haven't really kept up with MongoDB in a while. From looking at https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/, it's not completely clear to me when OP_MSG is actually used by the database. Do you know what it's used for? Basically, I'm trying to understand whether for this to be useful we need to parse additional information out of the message or for now we just want to know about the existence at all?

Reply:
Yeah, the quick description that I listed above isn't super clear, but the OP_MSG seems to be where the bulk of the information is now sent when going back and forth between mongo.  Here's a quick example, say we have this simple python script:

```python
from pymongo import MongoClient

def recreate():
    client = MongoClient('localhost', 10000)
    client.test.command("dropAllUsersFromDatabase")
    client.test.command("createUser", "shaun", pwd="password", roles=["read"])
    client.test.coll.drop()
    client.test.coll.insert_one({'name': 'foo', 'entry': 'bar'})

def main():
    # recreate()

    client = MongoClient('mongodb://shaun:password@localhost:10000/test')

    for entry in client.test.coll.find():
        print(entry)

if __name__ == '__main__':
    main()
```

When this script is run, the envoy mongoproxy log will look like this:

```
{"time": "2018-11-14T17:06:35.841Z", "message": {"opcode": "OP_QUERY", "id": 16807, "response_to": 0, "flags": "0x0", "collection": "admin.$cmd", "skip": 0, "return": -1, "query": {"ismaster": 1, "client": {"driver": {"name": "PyMongo", "version": "3.7.2"}, "os": {"type": "Darwin", "name": "Darwin", "architecture": "x86_64", "version": "10.13.6"}, "platform": "CPython 3.6.6.final.0"}}, "fields": {}}, "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.843Z", "message": {"opcode": "OP_REPLY", "id": 472, "response_to": 16807, "flags": "0x8", "cursor": "0", "from": 0, "returned": 1, "documents": 1}, "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.845Z", "message": {"opcode": "OP_QUERY", "id": 282475249, "response_to": 0, "flags": "0x0", "collection": "admin.$cmd", "skip": 0, "return": -1, "query": {"ismaster": 1, "client": {"driver": {"name": "PyMongo", "version": "3.7.2"}, "os": {"type": "Darwin", "name": "Darwin", "architecture": "x86_64", "version": "10.13.6"}, "platform": "CPython 3.6.6.final.0"}, "compression": {}}, "fields": {}}, "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.846Z", "message": {"opcode": "OP_REPLY", "id": 473, "response_to": 282475249, "flags": "0x8", "cursor": "0", "from": 0, "returned": 1, "documents": 1}, "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.847Z", "message": {"opcode": "OP_MSG", "id": 1622650073, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"ismaster": 1, "saslSupportedMechs": "test.shaun", "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.848Z", "message": {"opcode": "OP_MSG", "id": 474, "response_to": 1622650073, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"ismaster": true, "maxBsonObjectSize": 16777216, "maxMessageSizeBytes": 48000000, "maxWriteBatchSize": 100000, "localTime": 1542215195865, "logicalSessionTimeoutMinutes": 30, "minWireVersion": 0, "maxWireVersion": 7, "readOnly": false, "saslSupportedMechs": {"0": "SCRAM-SHA-1", "1": "SCRAM-SHA-256"}, "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.850Z", "message": {"opcode": "OP_MSG", "id": 984943658, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"saslStart": 1, "mechanism": "SCRAM-SHA-256", "payload": "n,,n=shaun,r=OTU0MDcwODExMDE5NDE1Nw==", "autoAuthorize": 1, "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.851Z", "message": {"opcode": "OP_MSG", "id": 475, "response_to": 984943658, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"conversationId": 1, "done": false, "payload": "r=OTU0MDcwODExMDE5NDE1Nw==MQQcFbLvSVPcJRzLMFAK3VtShkd1wbGR,s=OMdxp2Q38c1JgUao4dP4ELefgeq6VqzPJebAqg==,i=15000", "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.866Z", "message": {"opcode": "OP_MSG", "id": 1144108930, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"saslContinue": 1, "conversationId": 1, "payload": "c=biws,r=OTU0MDcwODExMDE5NDE1Nw==MQQcFbLvSVPcJRzLMFAK3VtShkd1wbGR,p=3yw49Tmw9PwTULK42g2F29Uu5NGu+wKJN7IIXN9hbP8=", "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.866Z", "message": {"opcode": "OP_MSG", "id": 476, "response_to": 1144108930, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"conversationId": 1, "done": false, "payload": "v=beOuCTEPlm/EqYVwKLlibKS2LkXECmqLlXo8nneMjco=", "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.867Z", "message": {"opcode": "OP_MSG", "id": 470211272, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"saslContinue": 1, "conversationId": 1, "payload": "", "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.868Z", "message": {"opcode": "OP_MSG", "id": 477, "response_to": 470211272, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"conversationId": 1, "done": true, "payload": "", "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.868Z", "message": {"opcode": "OP_MSG", "id": 101027544, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"find": "coll", "filter": {}, "lsid": {"id": "??.??Ex????v?2?"}, "$db": "test", "$readPreference": {"mode": "primaryPreferred"}}}]} , "upstream_host": "<ip_addr>:27017"}
{"time": "2018-11-14T17:06:35.869Z", "message": {"opcode": "OP_MSG", "id": 478, "response_to": 101027544, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"cursor": {"firstBatch": {"0": {"_id": "5bec5513b4815d794a04474a", "name": "foo", "entry": "bar"}}, "id": 0, "ns": "test.coll"}, "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
```

Without the patch, none of the OP_MSG queries will get written, so it's pretty difficult to figure out what was accessed in the database.

For our use case, just logging the OP_MSG to the mongoproxy log is sufficient.

Reply:
Huh, interesting. If they are moving to OP_MSG for all messages, eventually we are going to have to parse out the contents of OP_MSG for all the other stuff we are doing or I guess the filter is going to become useless? Can you potentially add some TODOs/comments around this somewhere?

Reply:
I'm not really sure what MongoDB's plan is moving forward, and I'm not really sure about all of your use cases, so it would be difficult for me to comment on that.

For our use case, simply logging the data that is transferred to and from MongoDB is a great start.  We'll probably also have to do some parsing of the logged data to extract out the bits of information that are useful to us.

Reply:
> Yeah, the quick description that I listed above isn't super clear, but the OP_MSG seems to be where the bulk of the information is now sent when going back and forth between mongo. Here's a quick example, say we have this simple python script:
> 
> ```python
> from pymongo import MongoClient
> 
> def recreate():
>     client = MongoClient('localhost', 10000)
>     client.test.command("dropAllUsersFromDatabase")
>     client.test.command("createUser", "shaun", pwd="password", roles=["read"])
>     client.test.coll.drop()
>     client.test.coll.insert_one({'name': 'foo', 'entry': 'bar'})
> 
> def main():
>     # recreate()
> 
>     client = MongoClient('mongodb://shaun:password@localhost:10000/test')
> 
>     for entry in client.test.coll.find():
>         print(entry)
> 
> if __name__ == '__main__':
>     main()
> ```
> 
> When this script is run, the envoy mongoproxy log will look like this:
> 
> ```
> {"time": "2018-11-14T17:06:35.841Z", "message": {"opcode": "OP_QUERY", "id": 16807, "response_to": 0, "flags": "0x0", "collection": "admin.$cmd", "skip": 0, "return": -1, "query": {"ismaster": 1, "client": {"driver": {"name": "PyMongo", "version": "3.7.2"}, "os": {"type": "Darwin", "name": "Darwin", "architecture": "x86_64", "version": "10.13.6"}, "platform": "CPython 3.6.6.final.0"}}, "fields": {}}, "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.843Z", "message": {"opcode": "OP_REPLY", "id": 472, "response_to": 16807, "flags": "0x8", "cursor": "0", "from": 0, "returned": 1, "documents": 1}, "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.845Z", "message": {"opcode": "OP_QUERY", "id": 282475249, "response_to": 0, "flags": "0x0", "collection": "admin.$cmd", "skip": 0, "return": -1, "query": {"ismaster": 1, "client": {"driver": {"name": "PyMongo", "version": "3.7.2"}, "os": {"type": "Darwin", "name": "Darwin", "architecture": "x86_64", "version": "10.13.6"}, "platform": "CPython 3.6.6.final.0"}, "compression": {}}, "fields": {}}, "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.846Z", "message": {"opcode": "OP_REPLY", "id": 473, "response_to": 282475249, "flags": "0x8", "cursor": "0", "from": 0, "returned": 1, "documents": 1}, "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.847Z", "message": {"opcode": "OP_MSG", "id": 1622650073, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"ismaster": 1, "saslSupportedMechs": "test.shaun", "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.848Z", "message": {"opcode": "OP_MSG", "id": 474, "response_to": 1622650073, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"ismaster": true, "maxBsonObjectSize": 16777216, "maxMessageSizeBytes": 48000000, "maxWriteBatchSize": 100000, "localTime": 1542215195865, "logicalSessionTimeoutMinutes": 30, "minWireVersion": 0, "maxWireVersion": 7, "readOnly": false, "saslSupportedMechs": {"0": "SCRAM-SHA-1", "1": "SCRAM-SHA-256"}, "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.850Z", "message": {"opcode": "OP_MSG", "id": 984943658, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"saslStart": 1, "mechanism": "SCRAM-SHA-256", "payload": "n,,n=shaun,r=OTU0MDcwODExMDE5NDE1Nw==", "autoAuthorize": 1, "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.851Z", "message": {"opcode": "OP_MSG", "id": 475, "response_to": 984943658, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"conversationId": 1, "done": false, "payload": "r=OTU0MDcwODExMDE5NDE1Nw==MQQcFbLvSVPcJRzLMFAK3VtShkd1wbGR,s=OMdxp2Q38c1JgUao4dP4ELefgeq6VqzPJebAqg==,i=15000", "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.866Z", "message": {"opcode": "OP_MSG", "id": 1144108930, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"saslContinue": 1, "conversationId": 1, "payload": "c=biws,r=OTU0MDcwODExMDE5NDE1Nw==MQQcFbLvSVPcJRzLMFAK3VtShkd1wbGR,p=3yw49Tmw9PwTULK42g2F29Uu5NGu+wKJN7IIXN9hbP8=", "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.866Z", "message": {"opcode": "OP_MSG", "id": 476, "response_to": 1144108930, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"conversationId": 1, "done": false, "payload": "v=beOuCTEPlm/EqYVwKLlibKS2LkXECmqLlXo8nneMjco=", "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.867Z", "message": {"opcode": "OP_MSG", "id": 470211272, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"saslContinue": 1, "conversationId": 1, "payload": "", "$db": "test", "$readPreference": {"mode": "primary"}}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.868Z", "message": {"opcode": "OP_MSG", "id": 477, "response_to": 470211272, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"conversationId": 1, "done": true, "payload": "", "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.868Z", "message": {"opcode": "OP_MSG", "id": 101027544, "response_to": 0, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"find": "coll", "filter": {}, "lsid": {"id": "??.??Ex????v?2?"}, "$db": "test", "$readPreference": {"mode": "primaryPreferred"}}}]} , "upstream_host": "<ip_addr>:27017"}
> {"time": "2018-11-14T17:06:35.869Z", "message": {"opcode": "OP_MSG", "id": 478, "response_to": 101027544, "flagBits": "0x0", "checksum": "0", "sections": [{"payloadType": 0, "payload": {"cursor": {"firstBatch": {"0": {"_id": "5bec5513b4815d794a04474a", "name": "foo", "entry": "bar"}}, "id": 0, "ns": "test.coll"}, "ok": 1.000000}}]} , "upstream_host": "<ip_addr>:27017"}
> ```
> Without the patch, none of the OP_MSG queries will get written, so it's pretty difficult to figure out what was accessed in the database.
> 
> For our use case, just logging the OP_MSG to the mongoproxy log is sufficient. @oneleggedredcow Can you please tell me how to enable logging to see what all queries are going to mongodb via istio-proxy ?

Reply:
Here's what the envoy.yaml file looks like for the example script given above:
```
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address:
      protocol: TCP
      address: 127.0.0.1
      port_value: 9901
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
    - filters:
      - name: envoy.mongo_proxy
        config:
          stat_prefix: mongo_logger
          access_log: "/tmp/mongo.log"
      - name: envoy.tcp_proxy
        config:
          stat_prefix: redirect_mongo
          cluster: service_local
  clusters:
  - name: service_local
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    # Comment out the following line to test on v6 networks
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    hosts:
      - socket_address:
          address: <ip_addr>
          port_value: 27017
```
Haven't played with istio-proxy yet, so I'm not sure how it translates, but that should get you pointed in the right direction.

Reply:
@oneleggedredcow thanks :) Have seen the similar in envoyproxy try scenarios . But some how it didn't work in istio or may needs additional steps for istio .